src/dataflow-types/src/types.rs

```diff
@@ -669,10 +669,18 @@ pub struct FileSourceConnector {
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
 pub struct S3SourceConnector {
     pub bucket: String,
+    pub key_sources: Vec<S3KeySource>,
     pub pattern: Option<Glob>,
     pub aws_info: aws::ConnectInfo,
 }
 
+/// A Source of Object Key names, the argument of the `OBJECTS FROM` clause
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum S3KeySource {
+    /// `OBJECTS FROM SCAN`
+    Scan,
+}
```

Comment:
I think here the docstrings should not mention SQL syntax, and should focus only on the semantics of what these things mean. The flip side of the discussion that we had here: https://github.com/MaterializeInc/materialize/pull/5194#discussion_r551560634